rfcs/0124-worker-manager.md

```diff
@@ -0,0 +1,271 @@
+components are intended to be developed as part of the worker manager codebase.
+
+## API
+The worker manager will expose APIs to control workers.  These include terminating
+instances and listing running workers.  There will also be APIs which operate
+on whole worker configurations and worker types.  These APIs will be standard
+Taskcluster APIs and will use scopes which are categorized by worker configuration
+and worker type.
+
+### Worker State
+Worker manager will also provide worker state APIs.  These APIs will show
+information like the last task runs for a given worker type, worker
+configuration or worker.  The data backing this information will be received
+from the Queue over RabbitMQ.  The worker manager will track outcomes.  The
+only information from the Queue stored in this service will be the task run
+outcome, task id and run id.  All other information will be available through
+the Queue.
+
+### Worker Configuration
+The worker manager will provide facilities for configuring Workers.  There will
+be a set of Taskcluster APIs for creating and managing worker configurations.
+Each worker configuration will specify global options like the maximum global
+capacity units, the scaling ratio and the list of worker types for which this
+worker configuration can be used.
+
+Worker configurations will also have per-providers options.  Each provider's
+options section will be able to override global options, like max capacity.
+This would be used to say "We want to have 1000 Capacity, but no more than 100
+from hardware".  Each provider a worker configuration can use would also have a
+section for provider-specific configuration.  This would store things like
+which worker image to use.  A more concrete example is that this section would
+store the AMI to use for an EC2 Worker.
+
+All levels of worker type configurations (e.g. global, per worker-type,
+per-provider) will support a `disabled: true` which will disable that section
+of the configuration.  The least specific disable will take priority.  This
+value will only block creation of new resources, not disable management of
+existing resources.
+
+Each edit of a worker-configuration will only be atomic to the
+worker-configuration level.
+
+### Multiple Worker Types for one Worker Configuration
+There are a lot of worker types which have identical configuration other than
+their worker type.  In order to support having more than one worker type for a
+given worker configuration, the worker configuration will have a list of worker
+types for which it can provide capacity for.  This will be a list in the
+configuration document.  Each entry in the list is either a string or an object
+which overrides global worker configuration options like max capacity for that
+specific worker type.  For configuration which is a range or value, the most
+restrictive will be used.
+
+Each worker type must be globally unique to a single worker configuration.
+
+### Documentation and Payload Schemata
```

Comment:
How will this get updated?

Reply:
I think I've added that info, but basically whoever/whatever manages each worker type should keep these things up to date.